lib/grandpa/vote_message.go

```diff
@@ -17,63 +16,33 @@ import (
 	"github.com/libp2p/go-libp2p-core/peer"
 )
 
+var errBeforeFinalizedBlock = errors.New("before latest finalized block")
+
 type networkVoteMessage struct {
 	from peer.ID
 	msg  *VoteMessage
 }
 
-// receiveVoteMessages receives messages from the in channel until a grandpa round finishes.
-func (s *Service) receiveVoteMessages(ctx context.Context) {
-	for {
-		select {
-		case msg, ok := <-s.in:
-			if !ok {
-				return
-			}
-
-			if msg == nil || msg.msg == nil {
-				continue
-			}
-
-			logger.Debugf("received vote message %v from %s", msg.msg, msg.from)
-			vm := msg.msg
-
-			switch vm.Message.Stage {
-			case prevote, primaryProposal:
-				s.telemetry.SendMessage(
-					telemetry.NewAfgReceivedPrevote(
-						vm.Message.BlockHash,
-						fmt.Sprint(vm.Message.Number),
-						vm.Message.AuthorityID.String(),
-					),
-				)
-			case precommit:
-				s.telemetry.SendMessage(
-					telemetry.NewAfgReceivedPrecommit(
-						vm.Message.BlockHash,
-						fmt.Sprint(vm.Message.Number),
-						vm.Message.AuthorityID.String(),
-					),
-				)
-			default:
-				logger.Warnf("unsupported stage %s", vm.Message.Stage.String())
-			}
-
-			v, err := s.validateVoteMessage(msg.from, vm)
-			if err != nil {
-				logger.Debugf("failed to validate vote message %v: %s", vm, err)
-				continue
-			}
-
-			logger.Debugf(
-				"validated vote message %v from %s, round %d, subround %d, "+
-					"prevote count %d, precommit count %d, votes needed %d",
-				v, vm.Message.AuthorityID, vm.Round, vm.Message.Stage,
-				s.lenVotes(prevote), s.lenVotes(precommit), s.state.threshold()+1)
-		case <-ctx.Done():
-			logger.Trace("returning from receiveMessages")
-			return
-		}
+func (s *Service) sendTelemetryVoteMessage(vm *VoteMessage) {
+	switch vm.Message.Stage {
+	case prevote, primaryProposal:
+		s.telemetry.SendMessage(
+			telemetry.NewAfgReceivedPrevote(
+				vm.Message.BlockHash,
+				fmt.Sprint(vm.Message.Number),
+				vm.Message.AuthorityID.String(),
+			),
+		)
+	case precommit:
+		s.telemetry.SendMessage(
+			telemetry.NewAfgReceivedPrecommit(
+				vm.Message.BlockHash,
+				fmt.Sprint(vm.Message.Number),
+				vm.Message.AuthorityID.String(),
+			),
+		)
+	default:
+		logger.Warnf("unsupported stage %s", vm.Message.Stage.String())
```

Comment:
nit no need to call `String()`, `%s` should fish it out itself

```suggestion
		logger.Warnf("unsupported stage %s", vm.Message.Stage.String())
```
```suggestion
		logger.Warnf("unsupported stage %s", vm.Message.Stage)
```